modules/backendscheduler/work/atomic.go

```diff
@@ -0,0 +1,44 @@
+package work
+
+import (
+	"os"
+	"path/filepath"
+)
+
+// atomicWriteFile writes data to a file atomically using a temporary file and rename.
+// This prevents race conditions and ensures readers never see partial data.
+func atomicWriteFile(data []byte, targetPath, tempPrefix string) error {
```

Comment:
What is concurrently reading the local work files?  It looks like the scheduler is always writing to them (as jobs are created and updated), and the flush to object storage is a remarshal from memory, and not reading the shard files.

Reply:
The message should say "readers never in a broken state".
The other condition I'd like to guard against is a crash right in the middle of a write.  If we take the atomic approach, I think we ensure that the work files on disk are always readable.